models/mli/fairness/model_fairxgb.py

```diff
@@ -0,0 +1,506 @@
+            colsample_bytree = [ 0.1*ii for ii in range(2,11,2)]
+            subsample = [1.0]
+            mu = [0.05*ii for ii in range(1, 14)]
+            
+        else:
+            eta= [0.1]
+            max_depth = list(range(4, 21, 2))
+            min_child_weight = [0.1, 0.5, 1.0, 2.0, 4.0, 8.0, 16.0, 32.0]
+            reg_lambda = [0.0, 0.1, 1.0, 5.0, 10.0]
+            colsample_bytree = [ 0.1*ii for ii in range(2,11,2)]
+            subsample = [1.0]
+            mu = [0.05*ii for ii in range(1, 14)]
+
+
+        self.params["eta"] = np.random.choice(eta)
+        self.params["max_depth"] = np.random.choice(max_depth)
+        self.params["min_child_weight"] = np.random.choice(min_child_weight)        
+        self.params["reg_lambda "] = np.random.choice(reg_lambda )        
+        self.params["colsample_bytree"] = np.random.choice(colsample_bytree) 
+        self.params["subsample"] = np.random.choice(subsample) 
+        self.params["mu"] = np.random.choice(mu) 
+
+    def _create_tmp_folder(self, logger):
+        # Create a temp folder to store files 
+        # Set the default value without context available (required to pass acceptance test)
+        tmp_folder = os.path.join(user_dir(), "%s_FAIRXGB_model_folder" % uuid.uuid4())
+        # Make a real tmp folder when experiment is available
+        if self.context and self.context.experiment_id:
+            tmp_folder = os.path.join(self.context.experiment_tmp_dir, "%s_FAIRXGB_model_folder" % uuid.uuid4())
+
+        # Now let's try to create that folder
+        try:
+            os.mkdir(tmp_folder)
+        except PermissionError:
+            # This not occur so log a warning
+            loggerwarning(logger, "FAIRXGB was denied temp folder creation rights")
+            tmp_folder = os.path.join(user_dir(), "%s_FAIRXGB_model_folder" % uuid.uuid4())
+            os.mkdir(tmp_folder)
+        except FileExistsError:
+            # We should never be here since temp dir name is expected to be unique
+            loggerwarning(logger, "FAIRXGB temp folder already exists")
+            tmp_folder = os.path.join(self.context.experiment_tmp_dir, "%s_FAIRXGB_model_folder" % uuid.uuid4())
+            os.mkdir(tmp_folder)
+        except:
+            # Revert to temporary file path
+            tmp_folder = os.path.join(user_dir(), "%s_FAIRXGB_model_folder" % uuid.uuid4())
+            os.mkdir(tmp_folder)
+
+        loggerinfo(logger, "FAIRXGB temp folder {}".format(tmp_folder))
+        return tmp_folder
+
+
+    def fit(self, X, y, sample_weight=None, eval_set=None, sample_weight_eval_set=None, **kwargs):
+             
+        # Specify these parameters for the dataset
+        # Also set feature engineering effort to 0
+        # under the features section of expert settings
+        ########################
+        # Specify the protected column
+        # Must be numeric
+        self.protected_name = "black"
```

Comment:
Another note is if there are several protected fields, not sure what could be done there, but worth thinking about.

Reply:
Yes, it could probably be extended to multiple protected values.  Right now, it's just using a single one as in the original paper.

Reply:
You can potentially combine protected fields to make a new column, e.g. gender and race in one column.